File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Source/Tva.Core/Tva.Measurements/ImmediateMeasurements.vb
'*******************************************************************************************************
'  Tva.Measurements.ImmediateMeasurements.vb - Collection of latest received measurements
'  Copyright © 2006 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: James R Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  This class represents the absolute latest received measurements
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  12/8/2005 - James R Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Namespace Measurements

    ' This is essentially a collection of measurements at a given timestamp
    Public Class ImmediateMeasurements

        Private m_measurements As Dictionary(Of Integer, TemporalMeasurement)
        Private m_taggedMeasurements As Dictionary(Of String, List(Of Integer))
        Private m_timeDeviationTolerance As Double
        Private m_ticks As Long

        Public Sub New(ByVal timeDeviationTolerance As Double)

            m_timeDeviationTolerance = timeDeviationTolerance
            m_measurements = New Dictionary(Of Integer, TemporalMeasurement)
            m_taggedMeasurements = New Dictionary(Of String, List(Of Integer))

        End Sub

        ''' <summary>Defines tagged measurements from a data table</summary>
        ''' <remarks>Expects tag to be aliased as "Tag" and measurement ID to be aliased as "ID"</remarks>
        Public Sub DefineTaggedMeasurements(ByVal measurements As DataTable)

            Dim tag As String
            Dim id As Integer

            For Each row As DataRow In measurements.Rows
                ' Get relevant fields
                tag = row("Tag")
                id = row("ID")

                ' Check for new tag
                If Not m_taggedMeasurements.ContainsKey(tag) Then
                    m_taggedMeasurements.Add(tag, New List(Of Integer))
                End If

                ' Add measurement to tag's measurement list
                m_taggedMeasurements(tag).Add(id)
            Next

        End Sub

        Public ReadOnly Property This() As ImmediateMeasurements
            Get
                Return Me
            End Get
        End Property

        Public Sub AddTaggedMeasurement(ByVal tag As String, ByVal measurementID As Integer)

            ' Check for new tag
            If Not m_taggedMeasurements.ContainsKey(tag) Then
                m_taggedMeasurements.Add(tag, New List(Of Integer))
            End If

            ' Add measurement to tag's measurement list
            m_taggedMeasurements(tag).Add(measurementID)

        End Sub

        Public Property Ticks() As Long
            Get
                Return m_ticks
            End Get
            Friend Set(ByVal value As Long)
                ' TODO: this must be equalized with SampleQueue base time - better to just reference that instead.
                m_ticks = value
            End Set
        End Property

        Public ReadOnly Property Timestamp() As Date
            Get
                Return New Date(m_ticks)
            End Get
        End Property

        Public ReadOnly Property MeasurementIDs() As Dictionary(Of Integer, TemporalMeasurement).KeyCollection
            Get
                Return m_measurements.Keys
            End Get
        End Property

        Public ReadOnly Property Tags() As Dictionary(Of String, List(Of Integer)).KeyCollection
            Get
                Return m_taggedMeasurements.Keys
            End Get
        End Property

        Public ReadOnly Property TagMeasurementIDs(ByVal tag As String) As List(Of Integer)
            Get
                Return m_taggedMeasurements(tag)
            End Get
        End Property

        Default Public ReadOnly Property Value(ByVal measurementID As Integer) As Double
            Get
                Return Measurement(measurementID)(m_ticks)
            End Get
        End Property

        Public ReadOnly Property Measurement(ByVal measurementID As Integer) As TemporalMeasurement
            Get
                SyncLock m_measurements
                    Dim value As TemporalMeasurement

                    If Not m_measurements.TryGetValue(measurementID, value) Then
                        ' Create new temporal measurement if it doesn't exist
                        value = New TemporalMeasurement(measurementID, Double.NaN, m_ticks, m_timeDeviationTolerance)
                        m_measurements.Add(measurementID, value)
                    End If

                    Return value
                End SyncLock
            End Get
        End Property

        Public Function CalculateAverage(ByRef count As Integer) As Double

            Dim measurement As Double
            Dim total As Double

            SyncLock m_measurements
                For Each entry As KeyValuePair(Of Integer, TemporalMeasurement) In m_measurements
                    measurement = Value(entry.Key)
                    If Not Double.IsNaN(measurement) Then
                        total += measurement
                        count += 1
                    End If
                Next
            End SyncLock

            Return total / count

        End Function

        Public Function CalculateTagAverage(ByVal tag As String, ByRef count As Integer) As Double

            Dim measurement As Double
            Dim total As Double

            For Each measurementID As Integer In m_taggedMeasurements(tag)
                measurement = Value(measurementID)
                If Not Double.IsNaN(measurement) Then
                    total += measurement
                    count += 1
                End If
            Next

            Return total / count

        End Function

        Public ReadOnly Property Minimum() As Double
            Get
                Dim minValue As Double = Double.MaxValue
                Dim measurement As Double

                SyncLock m_measurements
                    For Each entry As KeyValuePair(Of Integer, TemporalMeasurement) In m_measurements
                        measurement = Value(entry.Key)
                        If Not Double.IsNaN(measurement) Then
                            If measurement < minValue Then minValue = measurement
                        End If
                    Next
                End SyncLock

                Return minValue
            End Get
        End Property

        Public ReadOnly Property Maximum() As Double
            Get
                Dim maxValue As Double = Double.MinValue
                Dim measurement As Double

                SyncLock m_measurements
                    For Each entry As KeyValuePair(Of Integer, TemporalMeasurement) In m_measurements
                        measurement = Value(entry.Key)
                        If Not Double.IsNaN(measurement) Then
                            If measurement > maxValue Then maxValue = measurement
                        End If
                    Next
                End SyncLock

                Return maxValue
            End Get
        End Property

        Public ReadOnly Property TagMinimum(ByVal tag As String) As Double
            Get
                Dim minValue As Double = Double.MaxValue
                Dim measurement As Double

                For Each measurementID As Integer In m_taggedMeasurements(tag)
                    measurement = Value(measurementID)
                    If Not Double.IsNaN(measurement) Then
                        If measurement < minValue Then minValue = measurement
                    End If
                Next

                Return minValue
            End Get
        End Property

        Public ReadOnly Property TagMaximum(ByVal tag As String) As Double
            Get
                Dim maxValue As Double = Double.MinValue
                Dim measurement As Double

                For Each measurementID As Integer In m_taggedMeasurements(tag)
                    measurement = Value(measurementID)
                    If Not Double.IsNaN(measurement) Then
                        If measurement > maxValue Then maxValue = measurement
                    End If
                Next

                Return maxValue
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2008 TVA Code Library/Source/PCS.Core/Measurements/ImmediateMeasurements.cs
//*******************************************************************************************************
//  ImmediateMeasurements.cs
//  Copyright © 2008 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR 2W-C
//       Phone: 423/751-2827
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - J. Ritchie Carroll
//       Initial version of source generated for Super Phasor Data Concentrator.
//  02/23/2006 - J. Ritchie Carroll
//       Classes abstracted for general use and added to PCS code library.
//  09/17/2008 - James R Carroll
//       Converted to C#.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Data;

namespace PCS.Measurements
{
    /// <summary>Represents the absolute latest measurement values received by a <see cref="ConcentratorBase"/> implementation.</summary>
    public class ImmediateMeasurements : IDisposable
    {
        #region [ Members ]

        // Fields
        private ConcentratorBase m_parent;
        private Dictionary<MeasurementKey, TemporalMeasurement> m_measurements;
        private Dictionary<string, List<MeasurementKey>> m_taggedMeasurements;
        private bool m_disposed;

        #endregion

        #region [ Constructors ]

        internal ImmediateMeasurements(ConcentratorBase parent)
        {
            m_parent = parent;
            m_parent.LagTimeUpdated += m_parent_LagTimeUpdated;
            m_parent.LeadTimeUpdated += m_parent_LeadTimeUpdated;
            m_measurements = new Dictionary<MeasurementKey, TemporalMeasurement>();
            m_taggedMeasurements = new Dictionary<string, List<MeasurementKey>>();
        }

        /// <summary>
        /// Releases the unmanaged resources before the <see cref="ImmediateMeasurements"/> object is reclaimed by <see cref="GC"/>.
        /// </summary>
        ~ImmediateMeasurements()
        {
            Dispose(false);
        }

        #endregion

        #region [ Properties ]

        /// <summary>We retrieve adjusted measurement values within time tolerance of concentrator real-time.</summary>
        public double this[int measurementID, string source]
        {
            get
            {
                return this[new MeasurementKey(measurementID, source)];
            }
        }

        /// <summary>We retrieve adjusted measurement values within time tolerance of concentrator real-time.</summary>
        public double this[MeasurementKey key]
        {
            get
            {
                return Measurement(key).GetAdjustedValue(m_parent.RealTimeTicks);
            }
        }

        /// <summary>Returns key collection of measurement keys.</summary>
        public Dictionary<MeasurementKey, TemporalMeasurement>.KeyCollection MeasurementKeys
        {
            get
            {
                return m_measurements.Keys;
            }
        }

        /// <summary>Returns key collection for measurement tags.</summary>
        public Dictionary<string, List<MeasurementKey>>.KeyCollection Tags
        {
            get
            {
                return m_taggedMeasurements.Keys;
            }
        }

        /// <summary>Returns the minimum value of all measurements.</summary>
        /// <remarks>This is only useful if all measurements represent the same type of measurement.</remarks>
        public double Minimum
        {
            get
            {
                double minValue = double.MaxValue;
                double measurement;

                lock (m_measurements)
                {
                    foreach (MeasurementKey key in m_measurements.Keys)
                    {
                        measurement = this[key];
                        if (!double.IsNaN(measurement))
                        {
                            if (measurement < minValue)
                            {
                                minValue = measurement;
                            }
                        }
                    }
                }

                return minValue;
            }
        }

        /// <summary>Returns the maximum value of all measurements.</summary>
        /// <remarks>This is only useful if all measurements represent the same type of measurement.</remarks>
        public double Maximum
        {
            get
            {
                double maxValue = double.MinValue;
                double measurement;

                lock (m_measurements)
                {
                    foreach (MeasurementKey key in m_measurements.Keys)
                    {
                        measurement = this[key];
                        if (!double.IsNaN(measurement))
                        {
                            if (measurement > maxValue)
                            {
                                maxValue = measurement;
                            }
                        }
                    }
                }

                return maxValue;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Releases all the resources used by the <see cref="ImmediateMeasurements"/> object.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Releases the unmanaged resources used by the <see cref="ImmediateMeasurements"/> object and optionally releases the managed resources.
        /// </summary>
        /// <param name="disposing">true to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
        protected virtual void Dispose(bool disposing)
        {
            if (!m_disposed)
            {
                try
                {
                    if (disposing)
                    {
                        if (m_parent != null)
                        {
                            m_parent.LagTimeUpdated -= m_parent_LagTimeUpdated;
                            m_parent.LeadTimeUpdated -= m_parent_LeadTimeUpdated;
                        }
                        m_parent = null;

                        if (m_measurements != null)
                        {
                            m_measurements.Clear();
                        }
                        m_measurements = null;

                        if (m_taggedMeasurements != null)
                        {
                            m_taggedMeasurements.Clear();
                        }
                        m_taggedMeasurements = null;
                    }
                }
                finally
                {
                    m_disposed = true;  // Prevent duplicate dispose.
                }
            }
        }

        /// <summary>Returns measurement key list of specified tag, if it exists.</summary>
        public List<MeasurementKey> TagMeasurementKeys(string tag)
        {
            return m_taggedMeasurements[tag];
        }

        /// <summary>We only store a new measurement value that is newer than the cached value.</summary>
        internal void UpdateMeasurementValue(IMeasurement newMeasurement)
        {
            Measurement(newMeasurement.Key).SetValue(newMeasurement.Ticks, newMeasurement.Value);
        }

        /// <summary>Retrieves the specified immediate temporal measurement, creating it if needed.</summary>
        public TemporalMeasurement Measurement(int measurementID, string source)
        {
            return Measurement(new MeasurementKey(measurementID, source));
        }

        /// <summary>Retrieves the specified immediate temporal measurement, creating it if needed.</summary>
        public TemporalMeasurement Measurement(MeasurementKey key)
        {
            lock (m_measurements)
            {
                TemporalMeasurement value;

                if (!m_measurements.TryGetValue(key, out value))
                {
                    // Create new temporal measurement if it doesn't exist
                    value = new TemporalMeasurement(key.ID, key.Source, double.NaN, m_parent.RealTimeTicks, m_parent.LagTime, m_parent.LeadTime);
                    m_measurements.Add(key, value);
                }

                return value;
            }
        }

        /// <summary>Defines tagged measurements from a data table.</summary>
        /// <remarks>Expects tag field to be aliased as "Tag", measurement ID field to be aliased as "ID" and source field to be aliased as "Source".</remarks>
        public void DefineTaggedMeasurements(DataTable taggedMeasurements)
        {
            foreach (DataRow row in taggedMeasurements.Rows)
            {
                AddTaggedMeasurement(row["Tag"].ToString(), new MeasurementKey(System.Convert.ToInt32(row["ID"]), row["Source"].ToString()));
            }
        }

        /// <summary>Associates a new measurement ID with a tag, creating the new tag if needed.</summary>
        /// <remarks>Allows you to define "grouped" points so you can aggregate certain measurements.</remarks>
        public void AddTaggedMeasurement(string tag, MeasurementKey key)
        {
            // Check for new tag
            if (!m_taggedMeasurements.ContainsKey(tag))
                m_taggedMeasurements.Add(tag, new List<MeasurementKey>());

            // Add measurement to tag's measurement list
            List<MeasurementKey> measurements = m_taggedMeasurements[tag];

            if (measurements.BinarySearch(key) < 0)
            {
                measurements.Add(key);
                measurements.Sort();
            }
        }

        /// <summary>Calculates an average of all measurements.</summary>
        /// <remarks>This is only useful if all measurements represent the same type of measurement.</remarks>
        public double CalculateAverage(ref int count)
        {
            double measurement;
            double total = 0.0D;

            lock (m_measurements)
            {
                foreach (MeasurementKey key in m_measurements.Keys)
                {
                    measurement = this[key];
                    if (!double.IsNaN(measurement))
                    {
                        total += measurement;
                        count++;
                    }
                }
            }

            return total / count;
        }

        /// <summary>Calculates an average of all measurements associated with the specified tag.</summary>
        public double CalculateTagAverage(string tag, ref int count)
        {
            double measurement;
            double total = 0.0D;

            foreach (MeasurementKey key in m_taggedMeasurements[tag])
            {
                measurement = this[key];
                if (!double.IsNaN(measurement))
                {
                    total += measurement;
                    count++;
                }
            }

            return total / count;
        }

        /// <summary>Returns the minimum value of all measurements associated with the specified tag.</summary>
        public double TagMinimum(string tag)
        {
            double minValue = double.MaxValue;
            double measurement;

            foreach (MeasurementKey key in m_taggedMeasurements[tag])
            {
                measurement = this[key];
                if (!double.IsNaN(measurement))
                {
                    if (measurement < minValue)
                    {
                        minValue = measurement;
                    }
                }
            }

            return minValue;
        }

        /// <summary>Returns the maximum value of all measurements associated with the specified tag.</summary>
        public double TagMaximum(string tag)
        {
            double maxValue = double.MinValue;
            double measurement;

            foreach (MeasurementKey key in m_taggedMeasurements[tag])
            {
                measurement = this[key];
                if (!double.IsNaN(measurement))
                {
                    if (measurement > maxValue)
                    {
                        maxValue = measurement;
                    }
                }
            }

            return maxValue;
        }

        // We dyanmically respond to real-time changes in lead or lag time...
        private void m_parent_LagTimeUpdated(double lagTime)
        {
            lock (m_measurements)
            {
                foreach (MeasurementKey key in m_measurements.Keys)
                {
                    Measurement(key).LagTime = lagTime;
                }
            }
        }

        private void m_parent_LeadTimeUpdated(double leadTime)
        {
            lock (m_measurements)
            {
                foreach (MeasurementKey key in m_measurements.Keys)
                {
                    Measurement(key).LeadTime = leadTime;
                }
            }
        }

        #endregion
    }
}